src/client-side-encryption/autoEncrypter.ts

```diff
@@ -0,0 +1,665 @@
+     * Requires the MongoDB Crypt shared library, available in MongoDB 6.0 or higher.
+     */
+    cryptSharedLibPath?: string;
+    /**
+     * If specified, never use mongocryptd and instead fail when the MongoDB Crypt
+     * shared library could not be loaded.
+     *
+     * This is always true when `cryptSharedLibPath` is specified.
+     *
+     * Requires the MongoDB Crypt shared library, available in MongoDB 6.0 or higher.
+     */
+    cryptSharedLibRequired?: boolean;
+    /**
+     * Search paths for a MongoDB Crypt shared library to be used (instead of mongocryptd)
+     * Only for driver testing!
+     * @internal
+     */
+    cryptSharedLibSearchPaths?: string[];
+  };
+  proxyOptions?: ProxyOptions;
+  /** The TLS options to use connecting to the KMS provider */
+  tlsOptions?: CSFLEKMSTlsOptions;
+}
+
+/**
+ * @public
+ *
+ * Extra options related to the mongocryptd process
+ * \* _Available in MongoDB 6.0 or higher._
+ */
+export type AutoEncryptionExtraOptions = NonNullable<AutoEncryptionOptions['extraOptions']>;
+
+/** @public */
+export const AutoEncryptionLoggerLevel = Object.freeze({
+  FatalError: 0,
+  Error: 1,
+  Warning: 2,
+  Info: 3,
+  Trace: 4
+} as const);
+
+/**
+ * @public
+ * The level of severity of the log message
+ *
+ * | Value | Level |
+ * |-------|-------|
+ * | 0 | Fatal Error |
+ * | 1 | Error |
+ * | 2 | Warning |
+ * | 3 | Info |
+ * | 4 | Trace |
+ */
+export type AutoEncryptionLoggerLevel =
+  (typeof AutoEncryptionLoggerLevel)[keyof typeof AutoEncryptionLoggerLevel];
+
+// Typescript errors if we index objects with `Symbol.for(...)`, so
+// to avoid TS errors we pull them out into variables.  Then we can type
+// the objects (and class) that we expect to see them on and prevent TS
+// errors.
+/** @internal */
+const kDecorateResult = Symbol.for('@@mdb.decorateDecryptionResult');
+/** @internal */
+const kDecoratedKeys = Symbol.for('@@mdb.decryptedKeys');
+
+/**
+ * @internal An internal class to be used by the driver for auto encryption
+ * **NOTE**: Not meant to be instantiated directly, this is for internal use only.
+ */
+export class AutoEncrypter implements StateMachineExecutable {
+  _client: MongoClient;
+  _bypassEncryption: boolean;
+  _keyVaultNamespace: string;
+  _keyVaultClient: MongoClient;
+  _metaDataClient: MongoClient;
+  _proxyOptions: ProxyOptions;
+  _tlsOptions: CSFLEKMSTlsOptions;
+  _kmsProviders: KMSProviders;
+  _bypassMongocryptdAndCryptShared: boolean;
+  _contextCounter: number;
+
+  _mongocryptdManager?: MongocryptdManager;
+  _mongocryptdClient?: MongoClient;
+
+  /** @internal */
+  _mongocrypt: MongoCrypt;
+
+  /**
+   * Used by devtools to enable decorating decryption results.
+   *
+   * When set and enabled, `decrypt` will automatically recursively
+   * traverse a decrypted document and if a field has been decrypted,
+   * it will mark it as decrypted.  Compass uses this to determine which
+   * fields were decrypted.
+   */
+  [kDecorateResult]?: boolean;
```

Comment:
I don't feel strongly but a wise Bailey once said optional booleans make an enum with 3 values. Any harm in setting this to false?